### DISCUSSION TITLE
Use of Mfront with Comsol environement

### DISCUSSION THREAD

**Discussion by ATT2023** [thelfer/tfel#634] [Q&A]
Hello @thelfer , 

I'm Armel and a new user of Mfront. 

I need to use it in combinaison  with COMSOL for my PhD project. 

My project consist of developing new material model including viscoplasticity and damage in finite transformation, and coupled it with other physics using COMSOL. 

To start, I'm using Mfront on windows environment and I succeeded to compile and run the Norton model for Abaqus and Aster interfaces. I have also run the test with `mtest` and the result it's very nice.  

### **But when I try to compile the same code for COMSOL Interface I receive the following messages:**  

```
tenet@ DESKTOP-QIBS28B MINGW64 /c/Users/tenet/Desktop/CodeMfront/Nortoncomsol$ mfront --obuild --interface=comsol /c/Users/tenet/Desktop/CodeMfront/Nortoncomsol/nortoncomsol2.mfront
Treating target : all
comsolNorton.cxx: In function 'int Norton(const double*, double*, double*, const int*, const double*, const int*, double*, const int*, double*)':
comsolNorton.cxx:110:13: warning: ignoring return value of 'bool tfel::material::Norton<hypothesis, NumericType, false>::initialize() [with tfel::material::ModellingHypothesis::Hypothesis hypothesis = tfel::material::ModellingHypothesis::TRIDIMENSIONAL; NumericType = double]', declared with attribute 'nodiscard' [-Wunused-result]
  110 | b.initialize();
      | ~~~~~~~~~~~~^~
In file included from comsolNorton.cxx:11:
C:/Users/tenet/Desktop/CodeMfront/Nortoncomsol/nortoncomsol2.mfront:99:37: note: declared here
The following library has been built :
- libComsolBehaviour.dll :  Norton
```

### **When I tried to run the result with mtest, I obtained this message:** 

```
tenet@ DESKTOP-QIBS28B MINGW64 /c/Users/tenet/Desktop/CodeMfront/Nortoncomsol/src$ mtest /c/Users/tenet/Desktop/CodeMfront/Nortoncomsol/src/comsolnorton.mtest
terminate called after throwing an instance of 'std::runtime_error'
  what():  SingleStructureSchemeParser::SingleStructureScheme : error while parsing file 'C:/Users/tenet/Desktop/CodeMfront/Nortoncomsol/src/comsolnorton.mtest' at line '1'.
SingleStructureSchemeParser::handleBehaviour: unknown interface 'comsol'
```

**### Here is the mtest code:** 

```python
@ Behaviour<comsol> 'C:/Users/tenet/Desktop/CodeMfront/Nortoncomsol/src/libComsolBehaviour.dll' 'Norton';
@ MaterialProperty<constant> 'YoungModulus' 178600.0E6 ;
@ MaterialProperty<constant> 'PoissonRatio' 0.3 ;
@ MaterialProperty<constant> 'A' 8.e-67 ;
@ MaterialProperty<constant> 'm' 8.2 ;
@ ExternalStateVariable 'Temperature' 293.15 ;
@ ImposedStress 'SXX' { 0. :  0., 30. : 40.e6 };
@ ImposedStress 'SXY' { 0. :  0., 30. : 40.e6 };
@ Times { 0., 30. in 100 };
```

### **here is the mfront code:** 

```c++
@ Parser Implicit;
@ Behaviour Norton;
@ Algorithm NewtonRaphson_NumericalJacobian ;

@ RequireStiffnessTensor;

@ MaterialProperty real A;
@ MaterialProperty real m;

@ StateVariable real p ;

@ ComputeStress{
  sig = D*eel ;
}

@ Integrator{
  real seq = sigmaeq(sig) ;
  Stensor n = Stensor(0.) ;
  if(seq > 1.e-12){
    n = 1.5*deviator(sig)/seq ;
  }
  feel += dp*n-deto ;
  fp -= dt*A*pow(seq,m) ;
} // end of @ Integrator

@ TangentOperator{
  Stensor4 Je ;
  getPartialJacobianInvert(Je) ;
  Dt = D*Je ;
}
```

I'll very appreciate your help.

thanks in advance

**Comment by Algorithm** [comment]
Please remove @algorithm. This account is not associated with this project.

**Reply by ATT2023** [reply]
sorry for that, its part of Mfront code

**Comment by thelfer** [comment]
@ATT2023 Thanks for your interest.
To be clear, the `comsol` interface was developped but never really tested. The person that should have tested it changed his mind after I created the interface and, since I don't have access to `COMSOL`, it was left in an experimental state.
But if you are willing to make the tests, then we could work together and make it work.

**Reply by thelfer** [reply]
It should not be a lot of work: basically a set of traction/shear test on a unit cube in 3D and 2D.

**Reply by ATT2023** [reply]
I'm ready to collaborate. It's the core of my PhD project. 
here is my institutional email: armel.tene@ensta-paris.fr
It's up to you to plane how we will proceed. 
I hope that during the development we integrate the following points:
- integrate new strain-stress (second Piola-Kirchhoff & Green Lagrange) relation and Helmholtz free energy 
-  integrate the the multiplicative decomposition to account others physics
- see how to integrate external state variable ( coming from others physics like thermal, chemistry....)
-  account the dependence of mechanical parameters with respect to others state variables 

I'm totally free for that!! 
Thank in advance

**Reply by thelfer** [reply]
Great. I just contacted you by mail. I am currently on vacations, I won't be very responsive.

**Comment by tiliass** [comment]
Hello all, I'm very interested by your discussion. Did you manage to do the tests? Have a good day.
